### PR TITLE
CAMEL-9602 - ProducerTemplateBuilder

### PR DESCRIPTION
**Pull request by lburgazzoli**
This is a first cut of a ProducerTemplateBuilder which covers some basic cases.
It does not support exchange, processor and exchange patter.

Any feedback would be appreciated.

### CODE REVIEW

**Comment by lburgazzoli** [comment]
@davsclaus do you think this is good enough and matches your requirements ?

**Comment by davsclaus** [comment]
Yeah I wonder if you should have a method that returns the Exchange as result? But maybe you just use request(Exchange.class) as that variation?

Also if you want the Message as return, you use request(Message.class). Otherwise it would be the message body only being returned.

We should then also support this builder with dependency injection so you can do

```
@Autowited
ProducerTemplateBuilder template
```

And also with the `@Produces` etc
```
@Produces("activemq:queue:foo")
ProducerTemplateBuilder template
```

Then the template has been pre-configured to send to that uri.

But all that is a different ticket we can look at after having a builder.

**Comment by davsclaus** [comment]
And you may want to explain what that with/without event notifier means. We may want to remove that, eg keep the builder simple but for  the 95% use-cases.

**Comment by lburgazzoli** [comment]
I've amended the code so now:


a) advanced functionalities like events, cache, executors can be customized by providing a consumer

```java
ProducerTemplateBuilder.on(context)
    .withTemplateCustomizer(t -> t.setExecutorService(myExecutor))
    .withBody("the body")
    .to("direct:start")
    .request();
```

b) request can return an Exchange or a Message according to the type requested


Notes:
- the ProducerTemplate does not have an asyncRequest which returns an Exchnage so point b above is not applicable (but I may be wrong)
- I'm unsure about the name ProducerTemplateBuilder as a builder is supposed to "build" something here instead we are adding a DSL on top of  ProducerTemplate but well, I do not have any nicer name in my mind :-)
- should we provide also a clear method to cleanup headers/body ?

**Comment by davsclaus** [comment]
The asyncXXX methods is not much in use, so its okay not to have em.

Yeah the naming is always tricky. Its more a fluent producer style. `FluentProducerTemplate` ?

When do you want to use clear? What would the code/use-case be ?

**Comment by lburgazzoli** [comment]
+1 for FluentProducerTemplate

The usage of clear could to reuse the same fluent template i.e. you want so have a different set of headers. Maybe it is better to have a contextualized clear like clearHeaders(), clearBody() etc. I'm unsure about this, just thinking aloud.

**Comment by davsclaus** [comment]
Ah yeah such as in a for loop and you want to send to the same endpoint but change maybe the body/headers in the loop.

You could have the fluent dependency injected with the endpoint and other settings, and thus would like to reuse the instance but clear the headers/body in the loop.

But also we should try to keep the api small so the fluent do not have a ton of methods like the producer template got over time.

**Comment by lburgazzoli** [comment]
Or maybe add withHeaders(Map) and you can set it to null to clear it

```java
template.withHeaders(null)
        .withHeader("header1", "value")
        .withHeader("header2", "value")
```

Not very beautiful indeed.

**Comment by davsclaus** [comment]
Yeah setting null is not a good way. I suggest to go without the clear etc at first, and then we can see later if we need them.

**Comment by lburgazzoli** [comment]
- renamed to FluentProducerTemplate
- added clear headers/body
- the underlying template is now cached and reused among calls

Should we remove the sync* stuffs from the fluent variant?

**Comment by davsclaus** [comment]
Looks much better. 

The fluent builder has not so many methods so maybe its fine for now, we have until 2.18 is released to make up the api ;)

**Comment by lburgazzoli** [comment]
So ok to merge ?

**Comment by davsclaus** [comment]
+1

File: camel-core/src/main/java/org/apache/camel/builder/FluentProducerTemplate.java [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T request(Class<T> type) throws CamelExecutionException {
+        validate();
+
+        T result;
+        if (type == Exchange.class) {
+            result = (T)template().request(endpoint, processorSupplier.get());
+        } else if (type == Message.class) {
+            result = (T)template().request(endpoint, processorSupplier.get()).getIn();
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/builder/FluentProducerTemplate.java]: Ah here you may need to return OUT instead. eg do a hasOut ? getOut : getIn code on the exchange, so the correct Message instance is returned.

**Comment by lburgazzoli** [comment]
Is the package org.apache.camel.builder ok or better to move to org.apache.camel.impl ?

**Comment by lburgazzoli** [comment]
btw, send ha two variants in ProducerTemplate: 
- one throws an exception ```sendBody(ednpoint, body)``` 
- one does not ```send(endpoint, processor)```

Should we mimic that behavior in ```send()``` ?

**Comment by davsclaus** [comment]
I think builder is a good package, that is where the route builder is also.

**Comment by davsclaus** [comment]
If they return an exchange then they should not throw an exception, but set the exception on the exchange instead.

And if they throw an exception then that should be runtime based exceptions such as CamelExecutionException

**Comment by lburgazzoli** [comment]
I see, the difference is that sendBody is not supposed to return a value but in the fluent builder such possibility does not exists, should we add it ? I'd say no and one should eventually call ```ExchangeHelper.extractResultBody(result, null)``` , make sense ?

**Comment by davsclaus** [comment]
The methods are using this naming style

send = InOnly    (no response)
request = InOut    (response)

**Comment by lburgazzoli** [comment]
That's not what the ProducerTemplate does: send(endpoint, processor) does return but send Body just ignore the return result (it calls send under the hoods) and eventually throws and exception, what pattern should I follow  
for the fluent template?

**Comment by davsclaus** [comment]
We could also make send/request return the same. The difference is essentially if they set the MEP as InOnly or InOut.

And then they return by default the message body. So you need to specify (Exchange.class) if you want exchange returned. But that should be more seldom. As people often need only the body, or if not, then they can get the Message or Exchange by specifying those as the type.

Then its more consistent. And then send/request is only about setting the MEP.

**Comment by lburgazzoli** [comment]
I think I'll leave as it and we may change the APIs later on according on how people will use it.

**Pull request closed without merging**